### PR TITLE
encoding spec chars of uri path prameters

### PR DESCRIPTION
**Pull request by michalonderko**

### CODE REVIEW

File: router/Route.js [hunk 1]
```diff
@@ -341,11 +341,30 @@ export default class Route {
...
 	}
 
 	/**
+	 * Encode special characters of path parameter value.
+	 * 
+	 * @param {string|number} paramValue Path parameter value
+	 * @return {string} Encoded path parameter value.
+	 */
+	_encodePathParamValue(paramValue) {
+		// Percen sing (%) has to be first!
```
Comment (jurca) [inline router/Route.js]: Percent* sign*

File: router/Route.js [hunk 2]
```diff
@@ -341,11 +341,30 @@ export default class Route {
...
+		const charsToEncode = ['%', '/', '?', '#', ':'];
```
Comment (jurca) [inline router/Route.js]: Please turn this into a module-scoped constant `PARAMETER_CHARS_TO_ENCODE`

File: router/Route.js [hunk 3]
```diff
@@ -341,11 +341,30 @@ export default class Route {
...
+
+		let newParamValue = String(paramValue);
```
Comment (jurca) [inline router/Route.js]: 1. Calling `String` without `new` looks like an error at the first glance, that's why the linter complains as well :). Please use some other form of conversion, such as: `${paramValue}`.

2. `newParamValue` does not communicate the purpose of the variable, please use names such as `escapedParamValue` instead

File: router/Route.js [hunk 4]
```diff
@@ -341,11 +341,30 @@ export default class Route {
...
+
+		charsToEncode.forEach(char => {
+			newParamValue = newParamValue.replace(char, encodeURIComponent(char));
```
Comment (jurca) [inline router/Route.js]: 1. This is unnecessary repeated computation. Please add a module-scoped constant `PARAMETER_CHARS_TABLE`:

```javascript
PARAMETER_CHARS_TABLE = new Map(PARAMETER_CHARS_TO_ENCODE.map(char => [char, encodeURIComponent(char)]))
```

2. For-of is preferable in ima's code base to .forEach, please change this.

File: router/Route.js [hunk 5]
```diff
@@ -341,11 +341,30 @@ export default class Route {
...
 		);
 	}
 
 	/**
+	 * Encode special characters of path parameter value.
+	 * 
+	 * @param {string|number} paramValue Path parameter value
+	 * @return {string} Encoded path parameter value.
+	 */
+	_encodePathParamValue(paramValue) {
```
Comment (michalonderko) [inline router/Route.js]: @jurca It was request from our SEO master. After next discussion he changed his mind. So I am going to encode whole `paramValue`.

**Pull request merged**